File: src/main/java/li/cli/oc/OpenComputers.java
package li.cli.oc;

import li.cli.oc.items.Components;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder;
import net.minecraft.block.Blocks;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemStack;
import net.minecraft.util.Identifier;

public class OpenComputers implements ModInitializer {

	public static final String modId = "opencomputers";

	public static final ItemGroup ITEM_GROUP = FabricItemGroupBuilder.build(new Identifier(modId, "general"), () -> new ItemStack(Blocks.NETHERITE_BLOCK));

	@Override
	public void onInitialize() {
		Components.registerComponents();
	}
}

File: src/main/kotlin/li/cli/oc/OpenComputers.kt
package li.cli.oc

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import li.cli.oc.blockentity.commons.TecBlockEntity
import li.cli.oc.networking.ServerNetworkHandler
import li.cli.oc.utility.CraftingComponents
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerBlockEntityEvents
import java.util.*

object OpenComputers: ModInitializer {

	const val modId = "opencomputers"

	val ITEM_GROUP = FabricItemGroupBuilder.build(Identifier(modId, "general")) { ItemStack(Components.Blocks.CaseOne.block) }

	val GSON = GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.IDENTITY).setPrettyPrinting().create()

	override fun onInitialize() {
		ConfigLoader.initializeConfig()
		Components.registerComponents()
		CraftingComponents.registerSpecialCrafting()

		ServerBlockEntityEvents.BLOCK_ENTITY_LOAD.register { blockEntity, _ ->
			if(blockEntity !is TecBlockEntity) return@register;
			ServerNetworkHandler.registerToNetwork(blockEntity)
		}
		ServerBlockEntityEvents.BLOCK_ENTITY_UNLOAD.register { blockEntity, _ ->
			if(blockEntity !is TecBlockEntity) return@register;
			if(blockEntity.address != null) ServerNetworkHandler.unregisterNetwork(blockEntity.address!!)
		}

	}
}
